waku/v2/protocol/waku_lightpush/protocol.nim

```diff
@@ -0,0 +1,156 @@
+
+      let
+        pubSubTopic = rpc.request.pubSubTopic
+        message = rpc.request.message
+      debug "PushRequest", pubSubTopic=pubSubTopic, msg=message
+
+      var response: PushResponse
+      if not wl.relayReference.isNil():
+        let data = message.encode().buffer
+
+        # Assumimng success, should probably be extended to check for network, peers, etc
+        discard wl.relayReference.publish(pubSubTopic, data)
+        response = PushResponse(is_success: true, info: "Totally.")
+      else:
+        debug "No relay protocol present, unsuccesssful push"
+        response = PushResponse(is_success: false, info: "No relay protocol")
+        
+      
+      let rpc = PushRPC(requestId: rpc.requestId, response: response)
+      await conn.writeLp(rpc.encode().buffer)
+
+    if rpc.response != PushResponse():
+      waku_lightpush_messages.inc(labelValues = ["PushResponse"])
+      if rpc.response.isSuccess:
+        info "lightpush message success"
+      else:
+        info "lightpush message failure", info=rpc.response.info
+
+  wl.handler = handle
+  wl.codec = WakuLightPushCodec
+
+proc init*(T: type WakuLightPush, peerManager: PeerManager, rng: ref BrHmacDrbgContext, handler: PushRequestHandler, relay: WakuRelay = nil): T =
+  debug "init"
+  let rng = crypto.newRng()
+  let wl = WakuLightPush(rng: rng,
+                         peerManager: peerManager, 
+                         requestHandler: handler, 
+                         relayReference: relay)
+  wl.init()
+  
+  return wl
+
+
+proc setPeer*(wlp: WakuLightPush, peer: RemotePeerInfo) =
+  wlp.peerManager.addPeer(peer, WakuLightPushCodec)
+  waku_lightpush_peers.inc()
+
+
+proc request(wl: WakuLightPush, req: PushRequest, peer: RemotePeerInfo): Future[WakuLightPushResult[PushResponse]] {.async, gcsafe.} = 
+  let connOpt = await wl.peerManager.dialPeer(peer, WakuLightPushCodec)
+  if connOpt.isNone():
+    waku_lightpush_errors.inc(labelValues = [dialFailure])
+    return err(dialFailure)
+
+  let connection = connOpt.get()
+
+  let rpc = PushRPC(requestId: generateRequestId(wl.rng), request: req)
+  await connection.writeLP(rpc.encode().buffer)
+
+  var message = await connection.readLp(MaxRpcSize.int)
+  let res = PushRPC.init(message)
+
+  if res.isErr():
+    waku_lightpush_errors.inc(labelValues = [decodeRpcFailure])
+    return err(decodeRpcFailure)
+
+  let rpcRes = res.get()
+  if rpcRes.response == PushResponse():
+    return err("empty response body")
+
+  return ok(rpcRes.response)
+
+proc request*(wl: WakuLightPush, req: PushRequest): Future[WakuLightPushResult[PushResponse]] {.async, gcsafe.} =
+  let peerOpt = wl.peerManager.selectPeer(WakuLightPushCodec)
+  if peerOpt.isNone():
+    waku_lightpush_errors.inc(labelValues = [dialFailure])
+    return err(dialFailure)
+
+  return await wl.request(req, peerOpt.get())
+
+proc request*(wl: WakuLightPush, req: PushRequest, handler: PushResponseHandler) {.async, gcsafe,
```

Comment:
I think `lightpush*()` on the Nim API now uses the deprecated method [here](https://github.com/status-im/nwaku/blob/c67c4e13546e9c5489055f07638024de8ae3cd3a/waku/v2/node/wakunode2.nim#L364-L375). We can either maintain the existing `lightpush` method on the API (with callback) and just internally use the new `request*`, or also deprecate this method and provide a new `lightpush` API call without the callback. WDYT?

Reply:
It's a two-step process. First, refactor and mark as deprecated; then update all usages and remove the deprecated method.

I wanted to reduce the changes to the minimum amount of files to reduce the noise in the PR 😁

Reply:
Right. Sounds reasonable then to do it in a next PR. No real concerns. Feel free to go ahead and merge this.